sycl/doc/design/KernelFusionJIT.md

```diff
@@ -0,0 +1,173 @@
+
+As this is an LLVM based JIT, we can integrate it in various ways.
+
+### Kernel fusion pipeline
+
+The kernel fusion process is triggered by the `fuseKernels` function from the fusion JIT module.
+It takes as parameters the information required for the fusion: a context, a list of information about kernels to be fused, an ordered list of kernels to fuse (the same kernel may appear more than once), the name of the fused kernel, a list of identical parameters, and a list of buffers to promote to private or local memory and constant from the runtime to inject into the module.
+
+The function creates a new LLVM module that will contain the fused module and adds the kernels to fuse and their dependencies into this module.
+In the case the original module is SPIR-V, the module is first translated to LLVM IR and then added to the module.
+Information about the fusion is registered within the module (see section [Passing information to the fusion pipeline](#passing-information-to-the-fusion-pipeline)) and runs the fusion and optimization pipeline.
+
+The pipeline currently consists of the following passes (in order):
+
+  - `SYCLKernelFusion` performs the actual fusion process by inlining kernels to fuse inside the fused kernel
+  - Generic optimization passes: `IndVarSimplifyPass`, `LoopUnrollPass`, `SROAPass`, `InferAddressSpacesPass` to remove pointers to the generic address-space
+    - These optimizations are important to help the internalizer, see note below.
+  - `SYCLInternalizer` promotes buffer to local or private memory
+  - `SYCLCP` propagates runtime information as constant in the kernel
+  - Generic optimization passes post-fusion: `SROAPass`, `SCCPPass`, `InstCombinePass`, `SimplifyCFGPass`, `SROAPass`, `InstCombinePass`, `SimplifyCFGPass`, `ADCEPass`, `EarlyCSEPass`
+
+Note: ideally the `InferAddressSpacesPass` task should be done by the static compiler.
+However, to help the inference and (in fine) the internalizer passes in general, we need to run optimizations more aggressively than the static currently does.
+The main barrier that could prevent the internalization of buffers is the presence of generic pointers.
+As we need to change a pointer to generic to a pointer to private / local, if it is casted to generic we need to ensure that any casts back
+to global can be safely changed to the proper address space.
+A more precise inference pass and memory analysis will be required to lift this workaround.
+
+Once the pipeline is finished, the module is translated into a SPIR-V module and encapsulated into a `SYCLKernelInfo` object containing metadata required for its injection into the runtime.
+
+### Passing information to the fusion pipeline
+
+Most of the information passed from the runtime into the JIT is stored as metadata inside the LLVM module the JIT creates.
+This eases lit testing as all the information is self-contained in the module.
+The metadata is attached to a function that will become the fused kernel:
```

Comment:
```suggestion
This eases lit testing as all the information is self-contained in the module.

The metadata is attached to a function that will become the fused kernel:
```
Is this function emitted by SYCLKernelFusion pass?

Reply:
No, a stub kernel it is created at the same time as the module and metadata are attached to it. I updated the description in `Kernel fusion pipeline` to better reflect that.